[GLib] Lobotomize FileSystemGlib.cpp

mcatanzaro opened the pull request:
#### 448487f4e635fac0719e05793ab6fba164ae361a
<pre>
[GLib] Lobotomize FileSystemGlib.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=255932">https://bugs.webkit.org/show_bug.cgi?id=255932</a>

Reviewed by Carlos Garcia Campos.

Most of FileSystemGlib.cpp is redundant with FileSystemPOSIX.cpp. There
is no good reason to implement the POSIX-style WTF::Filesystem APIs
using GIO when we have access to POSIX. FileSystemGlib.cpp nowadays goes
to rather extreme lengths to work around impedance between the GIO and
POSIX APIs, and it&apos;s much simpler to just use POSIX.

* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/PlatformGTK.cmake:
* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/glib/FileSystemGlib.cpp:
(WTF::FileSystemImpl::stringFromFileSystemRepresentation): Deleted.
(WTF::FileSystemImpl::fileSystemRepresentation): Deleted.
(WTF::FileSystemImpl::genericGIOFileClose): Deleted.
(WTF::FileSystemImpl::genericGIOFileQueryInfo): Deleted.
(WTF::FileSystemImpl::genericGIOGetInputStream): Deleted.
(WTF::FileSystemImpl::genericGIOGetOutputStream): Deleted.
(WTF::FileSystemImpl::genericGIOGetFileDescriptorBased): Deleted.
(WTF::FileSystemImpl::posixFileDescriptor): Deleted.
(WTF::FileSystemImpl::fileSize): Deleted.
(WTF::FileSystemImpl::fileID): Deleted.
(WTF::FileSystemImpl::fileIDsAreEqual): Deleted.
(WTF::FileSystemImpl::fileCreationTime): Deleted.
(WTF::FileSystemImpl::openTemporaryFile): Deleted.
(WTF::FileSystemImpl::openFile): Deleted.
(WTF::FileSystemImpl::closeFile): Deleted.
(WTF::FileSystemImpl::seekFile): Deleted.
(WTF::FileSystemImpl::truncateFile): Deleted.
(WTF::FileSystemImpl::flushFile): Deleted.
(WTF::FileSystemImpl::writeToFile): Deleted.
(WTF::FileSystemImpl::readFromFile): Deleted.
(WTF::FileSystemImpl::getFileDeviceId): Deleted.
(WTF::FileSystemImpl::volumeFileBlockSize): Deleted.
(WTF::FileSystemImpl::lockFile): Deleted.
(WTF::FileSystemImpl::unlockFile): Deleted.
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::fileCreationTime):
(WTF::FileSystemImpl::temporaryFileDirectory):
(WTF::FileSystemImpl::openTemporaryFile):
* Source/cmake/OptionsCommon.cmake:

Canonical link: <a href="https://commits.webkit.org/263503@main">https://commits.webkit.org/263503@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/59e93d3ad8f5d448730b4d6cdea021f82d1e1be2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4802 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4923 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5091 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6316 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4930 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5084 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4895 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/5176 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4883 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4966 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4305 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6329 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/2454 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4300 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/9273 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/3989 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4316 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4375 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/5956 "Passed tests") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/4552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4780 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/3896 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/4899 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4287 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1201 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1177 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8350 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/5030 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/4650 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1314 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WTF/wtf/glib/FileSystemGlib.cpp:
```diff
@@ -125,290 +59,6 @@ String filenameForDisplay(const String& string)
-    return a == b;
-}
-
-std::optional<WallTime> fileCreationTime(const String& path)
-{
-    const auto filename = fileSystemRepresentation(path);
-    if (!validRepresentation(filename))
-        return std::nullopt;
-
-    GRefPtr<GFile> file = adoptGRef(g_file_new_for_path(filename.data()));
-    GRefPtr<GFileInfo> info = adoptGRef(g_file_query_info(file.get(), G_FILE_ATTRIBUTE_TIME_CREATED, G_FILE_QUERY_INFO_NONE, nullptr, nullptr));
-    if (info && g_file_info_has_attribute(info.get(), G_FILE_ATTRIBUTE_TIME_CREATED))
-        return WallTime::fromRawSeconds(g_file_info_get_attribute_uint64(info.get(), G_FILE_ATTRIBUTE_TIME_CREATED));
-
-    return std::nullopt;
```
We can't remove this one, at least not without adding an implementation in FileSystemPOSIX.cpp

mcatanzaro replied:
OK, I see there's no Linux implementation in FileSystemPOSIX.cpp. We can use statx for this.

carlosgcampos commented on Source/WTF/wtf/glib/FileSystemGlib.cpp:
```diff
@@ -125,290 +59,6 @@ String filenameForDisplay(const String& string)
-    return std::nullopt;
-}
-
-String openTemporaryFile(StringView prefix, PlatformFileHandle& handle, StringView suffix)
-{
-    // FIXME: Suffix is not supported, but OK for now since the code using it is macOS-port-only.
-    ASSERT_UNUSED(suffix, suffix.isEmpty());
-
-    GUniquePtr<gchar> filename(g_strdup_printf("%s%s", prefix.utf8().data(), createVersion4UUIDString().utf8().data()));
-    GUniquePtr<gchar> tempPath(g_build_filename(g_get_tmp_dir(), filename.get(), nullptr));
-    GRefPtr<GFile> file = adoptGRef(g_file_new_for_path(tempPath.get()));
-
-    handle = G_SEEKABLE(g_file_create_readwrite(file.get(), G_FILE_CREATE_NONE, nullptr, nullptr));
-    if (!isHandleValid(handle))
-        return String();
-    return String::fromUTF8(tempPath.get());
-}
```
I'm not sure about this one either, because g_get_temp_dir() is not exactly the same as what openTemporaryFile does. The main difference is that P_tmpdir is used before falling back to /tmp. I'm not sure it's important though. Maybe we can extract that logic into FileSystem::tempDir or something like that.

mcatanzaro replied:
OK, let's keep using g_get_temp_dir().

carlosgcampos reviewed (changes requested):
Note also that glib uses statx in Linux if available which is supposed to be faster than stat. Maybe we want to add FileSystemLinux.cpp and use statx if available too. I also wonder about the FILESYSTEM_POSIX_FAST_PATH, is std::filesystem slower in other platforms too?

carlosgcampos commented on Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
```diff
@@ -177,22 +182,31 @@ std::optional<uint64_t> fileSize(PlatformFileHandle handle)
     if (!fsRep.data() || fsRep.data()[0] == '\0')
         return std::nullopt;
 
+#if OS(LINUX)
```
We probably need to check statx is available at build time

mcatanzaro replied:
OK, done.

carlosgcampos commented on Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
```diff
@@ -177,22 +182,30 @@ std::optional<uint64_t> fileSize(PlatformFileHandle handle)
 std::optional<WallTime> fileCreationTime(const String& path)
 {
-#if OS(DARWIN) || OS(OPENBSD) || OS(NETBSD) || OS(FREEBSD)
+#if OS(LINUX) && HAVE(STATX) || OS(DARWIN) || OS(OPENBSD) || OS(NETBSD) || OS(FREEBSD)
```
Don't we need parentheses? 
```suggestion
#if (OS(LINUX) && HAVE(STATX)) || OS(DARWIN) || OS(OPENBSD) || OS(NETBSD) || OS(FREEBSD)
```

mcatanzaro replied:
I don't think the parentheses are required, but they certainly won't hurt either.

mcatanzaro committed:
[GLib] Lobotomize FileSystemGlib.cpp
https://bugs.webkit.org/show_bug.cgi?id=255932

Reviewed by Carlos Garcia Campos.

Most of FileSystemGlib.cpp is redundant with FileSystemPOSIX.cpp. There
is no good reason to implement the POSIX-style WTF::Filesystem APIs
using GIO when we have access to POSIX. FileSystemGlib.cpp nowadays goes
to rather extreme lengths to work around impedance between the GIO and
POSIX APIs, and it's much simpler to just use POSIX.

* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/PlatformGTK.cmake:
* Source/WTF/wtf/PlatformWPE.cmake:
* Source/WTF/wtf/glib/FileSystemGlib.cpp:
(WTF::FileSystemImpl::stringFromFileSystemRepresentation): Deleted.
(WTF::FileSystemImpl::fileSystemRepresentation): Deleted.
(WTF::FileSystemImpl::genericGIOFileClose): Deleted.
(WTF::FileSystemImpl::genericGIOFileQueryInfo): Deleted.
(WTF::FileSystemImpl::genericGIOGetInputStream): Deleted.
(WTF::FileSystemImpl::genericGIOGetOutputStream): Deleted.
(WTF::FileSystemImpl::genericGIOGetFileDescriptorBased): Deleted.
(WTF::FileSystemImpl::posixFileDescriptor): Deleted.
(WTF::FileSystemImpl::fileSize): Deleted.
(WTF::FileSystemImpl::fileID): Deleted.
(WTF::FileSystemImpl::fileIDsAreEqual): Deleted.
(WTF::FileSystemImpl::fileCreationTime): Deleted.
(WTF::FileSystemImpl::openTemporaryFile): Deleted.
(WTF::FileSystemImpl::openFile): Deleted.
(WTF::FileSystemImpl::closeFile): Deleted.
(WTF::FileSystemImpl::seekFile): Deleted.
(WTF::FileSystemImpl::truncateFile): Deleted.
(WTF::FileSystemImpl::flushFile): Deleted.
(WTF::FileSystemImpl::writeToFile): Deleted.
(WTF::FileSystemImpl::readFromFile): Deleted.
(WTF::FileSystemImpl::getFileDeviceId): Deleted.
(WTF::FileSystemImpl::volumeFileBlockSize): Deleted.
(WTF::FileSystemImpl::lockFile): Deleted.
(WTF::FileSystemImpl::unlockFile): Deleted.
* Source/WTF/wtf/posix/FileSystemPOSIX.cpp:
(WTF::FileSystemImpl::fileCreationTime):
(WTF::FileSystemImpl::temporaryFileDirectory):
(WTF::FileSystemImpl::openTemporaryFile):
* Source/cmake/OptionsCommon.cmake:

Canonical link: https://commits.webkit.org/263503@main
